File: 022024/02 - api/python/api/fipe/service/fipe_service.py
# fipe/services/fipe_service.py
import requests

class FipeService:
    def consultar_url(self, api_url):
        response = requests.get(api_url)
        if response.status_code == 200:
            return response.json()
        else:
            return {"error": f"Falha ao obter dados. Código de status: {response.status_code}"}

    def consultar_marcas(self):
        return self.consultar_url("https://parallelum.com.br/fipe/api/v1/carros/marcas")

    def consultar_modelos(self, id):
        return self.consultar_url(f"https://parallelum.com.br/fipe/api/v1/carros/marcas/{id}/modelos")

    def consultar_anos(self, marca, modelo):
        return self.consultar_url(f"https://parallelum.com.br/fipe/api/v1/carros/marcas/{marca}/modelos/{modelo}/anos")

    def consultar_valor(self, marca, modelo, ano):
        return self.consultar_url(f"https://parallelum.com.br/fipe/api/v1/carros/marcas/{marca}/modelos/{modelo}/anos/{ano}")

File: ConstrucaoServicoWeb/exemplo02/fipe/src/main/java/com/api/fipe/service/FipeService.java
package com.api.fipe.service;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

@Service
public class FipeService {

    @Value("${fipe.api.base-url:https://parallelum.com.br/fipe/api/v1}")
    private String baseUrl;

    private String consultarURL(String apiUrl) {
        RestTemplate restTemplate = new RestTemplate();
        ResponseEntity<String> responseEntity = restTemplate.getForEntity(apiUrl, String.class);
        if (responseEntity.getStatusCode().is2xxSuccessful()) {
            return responseEntity.getBody();
        } else {
            return "Falha ao obter dados. Código de status: " + responseEntity.getStatusCode();
        }
    }

    public String consultarMarcas() {
        return consultarURL(baseUrl + "/carros/marcas");
    }

    public String consultarModelos(int id) {
        return consultarURL(baseUrl + "/carros/marcas/" + id + "/modelos");
    }

    public String consultarAnos(int marca, int modelo) {
        return consultarURL(baseUrl + "/carros/marcas/" + marca + "/modelos/" + modelo + "/anos");
    }

    public String consultarValor(int marca, int modelo, String ano) {
        return consultarURL(baseUrl + "/carros/marcas/" + marca + "/modelos/" + modelo + "/anos/" + ano);
    }
}
